Add EIP-5976: Handbook for EIP Authors (#5976)
ethereum/EIPs: EIPS/eip-5976.md

> ---

Author: Rationale and Security Considerations were suggested to be removed by Editor. This will need a manual merge

Reviewer: You should add this to:

https://github.com/ethereum/EIPs/blob/cf35f4c7097dd12618a741c5eb69199e0c72e6df/.github/workflows/ci.yml#L119

status: Draft
type: Informational
created: 2022-11-15
---

## Abstract

A handy reference for EIP authors with specific suggestions of best practices for different kinds of EIPs.

## Specification

### Scoping

1. Standard track EIPs are divided in 4 categories: Core, Network, Interface, and ERC (Application). A new EIP should only address one category, although ERCs may occasionally make Interface-category changes (a good example is [EIP-3668](./eip-3668.md).

### Identifying Related and Competing EIPs

1. Authors of the New EIP SHOULD familiarize with related EIPs.
2. For existing EIPs that the New EIP strictly depends on, Author SHOULD mention them in its Specification section and and `requires` field in the EIP preamble.
3. For EIPs that could depend on or interact with the new EIP, Author is
RECOMMENDED to properly describe how these EIPs could interact with new EIP.
4. For EIPs that the new EIP _competes with_, Authors SHOULD describe in Motivation section how the new EIP addresses new areas of use cases old ones don't, and in
Rationale section that how the design decisions in the new EIP compare with
the existing competing EIPs.

> ### Finalization

Reviewer 2: ```suggestion
### The EIP Process
```

> 1. New EIPs are RECOMMENDED to turn into FINAL when two or more versions of independent and different compliant implementations are built and deployed.

Reviewer 2: I'd like this to be a bit more comprehensive:
```suggestion
1. EIPs should move to Review once a significant amount of discussion has occurred and consensus has been reached, either by edits to the EIP or the creation of a new competing EIP. It is recommended that this discussion occur on FEM and/or be linked in the FEM discussion.
2. Non-core EIPs should move to last call when two or more independently-built compliant implementations have been created. Core EIPs should move to last call when they are proposed for a fork. Make sure to have it thoroughly proofread by at least one author and one non-author, as major mistakes (as with [EIP-712](./eip-712.md) and [EIP-1363](./eip-1363.md)) might not be possible to correct once the EIP moves to final.
3. Non-core EIPs should move to final as soon as their last call deadline is reached, as long as there remains no contentious discssion.
```

---
eip: 5976
title: EIP Author Handbook
description: Handy reference for EIP authors with specific suggestions of best practice
author: Zainan Victor Zhou (@xinbenlv)
discussions-to: https://ethereum-magicians.org/t/informational-eip-author-handbook/11754
status: Draft
type: Informational
created: 2022-11-15
---

## Abstract

A handy reference for EIP authors with specific suggestions of best practices for different kinds of EIPs.

## Specification

> 

Reviewer 2: ```suggestion
```

Rationale section that how the design decisions in the new EIP compare with
the existing competing EIPs.

### Finalization

1. New EIPs are RECOMMENDED to turn into FINAL when two or more versions of independent and different compliant implementations are built and deployed.

### ERC EIPs

#### Worthiness of ERC EIPs

1. New ERC EIPs is RECOMMENDED to justify its "worthiness" by providing in its Rationale section:

- A _massive pre-standardized adoptions_, e.g [EIP-721 Non-Fungible Token](./eip-721.md).
- A _strong potential demand as seen elsewhere_, but couldn't yet be achieved on EVM due to lacking API consensus [EIP-137 Ethereum Domain Name Service](./eip-137.md).
- A _novel use cases backed_ by interested future adopters., e.g. [EIP-4337](./eip-4337.md).

### Core EIPs

#### When requiring a hard-fork

1. A new EIP requiring a hard-fork is RECOMMENDED to explicitly declare a place-holding number as a named parameter and recommend implementing clients to use such name as controlling parameter prior to rollout.

#### When adding a new EVM instruction

> For a new EIP that adds EVM Instruction, here is a handy check-list

Reviewer 2: ```suggestion
For a new EIP that adds a new EVM Opcode/Instruction:
```

> - Its Category MUST be `Core`

Reviewer 2: ```suggestion
- Its `type` MUST be `Standards Track` and its `category` MUST be `Core`.
```

> - It MUST specify a mnemonic

Reviewer 2: ```suggestion
- It MUST specify a mnemonic (such as `BASEFEE` for [EIP-3198](./eip-3198.md)).
```

> - It MUST specify a opcode number. It should contains wording this opcode number are deemed subject to change. Do not use a placeholder opcode number so clients of different language can integrate before. Author SHOULD check other instruction-adding non-finalized EIPs to avoid picking a conflicting opcode.

Reviewer 2: ```suggestion
- It MUST specify a placeholder opcode number, which MUST be clearly indicated as subject to change until the EIP nears finalization. When reviewing the opcode number, authors MUST check other instruction-adding EIPs to avoid picking a conflicting opcode.
```

> before finalize.

Reviewer 2: ```suggestion
```

> - It MUST specify a gas cost. It MUST justify its gas cost choice in is Rationale section.

Reviewer 2: ```suggestion
- It MUST specify a gas cost. It MUST justify its gas cost choice in its Rationale section.
```

## Rationale

> Help authors better streamline the authoring and reduce review round-trip time.

Reviewer 2: Can also be omitted.

Author: I think this is legit rationale and I lean towards keeping the Rationele unless you are against...

Reviewer 2: I am against including a rationale for this EIP, as:

1. This is actually motivation, not Rationale
2. The "design decisions" made in this EIP are exceptionally obvious and well-documented in EIP-1.

Author: Got it. Removed, respect your opposition. Pinging @SamWilsn for updating EIPW to exempt "Informational" track from mandating "Rationale " and "Security Considerations"

- A _novel use cases backed_ by interested future adopters., e.g. [EIP-4337](./eip-4337.md).

### Core EIPs

#### When requiring a hard-fork

1. A new EIP requiring a hard-fork is RECOMMENDED to explicitly declare a place-holding number as a named parameter and recommend implementing clients to use such name as controlling parameter prior to rollout.

#### When adding a new EVM instruction

For a new EIP that adds EVM Instruction, here is a handy check-list

- Its Category MUST be `Core`
- It MUST specify a mnemonic
- It MUST specify a opcode number. It should contains wording this opcode number are deemed subject to change. Do not use a placeholder opcode number so clients of different language can integrate before. Author SHOULD check other instruction-adding non-finalized EIPs to avoid picking a conflicting opcode.
before finalize.
- It MUST specify a gas cost. It MUST justify its gas cost choice in is Rationale section.

## Rationale

Help authors better streamline the authoring and reduce review round-trip time.

## Security Considerations

> No security issues introduced.

Reviewer 2: Can be omitted for informational EIPs.

Reviewer 2: Can be overwritten.

Author: Should we remove that from EIPW side?

Reviewer 2: Yes, it should. CC @SamWilsn.
